application/comit_node/src/connection_pool.rs

```diff
@@ -0,0 +1,286 @@
+use crate::{
+    bam_api::rfc003::swap_config, node_id::NodeId, swap_protocols::rfc003::bob::BobSpawner,
+};
+use bam::{self, json};
+use futures::{
+    future::{self, Shared},
+    Future,
+};
+use std::{
+    collections::HashMap,
+    fmt::{self, Debug},
+    io,
+    net::SocketAddr,
+    sync::{Arc, Mutex, RwLock},
+};
+use tokio::{
+    self,
+    io::{AsyncRead, AsyncWrite},
+    net::TcpStream,
+};
+
+type BamClient =
+    Arc<Mutex<bam::client::Client<json::Frame, json::OutgoingRequest, json::Response>>>;
+
+enum ConnectionState {
+    Connected {
+        client: BamClient,
+    },
+    Connecting {
+        waiting_list: Shared<Box<dyn Future<Item = BamClient, Error = io::ErrorKind> + Send>>,
+    },
+    Disconnected,
+}
+
+impl Debug for ConnectionState {
+    fn fmt<'a>(&self, f: &mut fmt::Formatter<'a>) -> Result<(), fmt::Error> {
+        write!(f, "{}", &self)
+    }
+}
+
+impl fmt::Display for ConnectionState {
+    fn fmt<'a>(&self, f: &mut fmt::Formatter<'a>) -> Result<(), fmt::Error> {
+        use self::ConnectionState::*;
+        match self {
+            Connected { .. } => write!(f, "connected"),
+            Connecting { .. } => write!(f, "connecting"),
+            Disconnected => write!(f, "disconnected"),
+        }
+    }
+}
+
+/// Keeps track of connections per NodeId.
+/// For now we don't allow multiple connections from the same NodeId.
+#[derive(Default, Debug)]
+pub struct ConnectionPool {
```

Comment:
:x: This guy has *A LOT* of logic but not a single test :(

I'd be interesting in doing some pair-programming on that since it seems to be quite a complex and important piece :)

Reply:
The logic and the api will change soon so I'm reluctant to test this at all atm.

I'd suggest pair programming the right logic with TDD.